src/json.hpp

```diff
@@ -1397,120 +1397,69 @@ constexpr T static_const<T>::value;
 /// abstract input adapter interface
 struct input_adapter_protocol
 {
-    virtual int get_character() = 0;
-    virtual std::string read(std::size_t offset, std::size_t length) = 0;
+    virtual int get_character() = 0; // returns characters in range [0,255], or eof() (a -'ve value)
```

Comment:
Just a question: what do you mean with `(a -'ve value)`?

Reply:
`(a negative value)`